File: ObjC/RestaurantWidget/Common/MenuItemsManager.m
//
//  DataManager.m
//  RestaurantWidget
//
//  Created by AppsFoundation on 23/09/15.
//  Copyright (c) 2015 AppsFoundation. All rights reserved.
//

#import "MenuItemsManager.h"
#import "MenuItem.h"

@implementation MenuItemsManager

#pragma mark - Singleton method

+ (MenuItemsManager *)sharedManager {
    static dispatch_once_t predicate = 0;
    static MenuItemsManager *sharedObject;

    dispatch_once(&predicate, ^{
        sharedObject = [[self alloc] init];
    });
    return sharedObject;
}

#pragma mark - Public methods

- (NSArray *)loadData {
    NSString* path = [[NSBundle mainBundle] pathForResource:@"MenuItems" ofType:@"plist"];
    NSDictionary *dataDictionary = [[NSDictionary alloc] initWithContentsOfFile:path];
    
    return [self constructDataFromDict:dataDictionary];
}

#pragma mark - Private methods

- (NSArray *)constructDataFromDict:(NSDictionary *)dict {
    
    NSMutableArray *resultArray = [NSMutableArray array];
    
    for (NSDictionary *obj in [dict allValues]) {
        MenuItem *loadedData = [[MenuItem alloc] init];
        loadedData.name = obj[@"name"];
        loadedData.image = obj[@"image"];
        loadedData.price = obj[@"price"];
        loadedData.ingredients = obj[@"ingredients"];
        loadedData.discount = obj[@"discount"];
        [resultArray addObject:loadedData];
    }
    
    return [NSArray arrayWithArray:resultArray];
}

@end

File: Swift/Restaurant/Common/MenuItemsManager.swift
//
//  MenuItemsManager.swift
//  Restaurant
//
//  Created by AppsFoundation on 8/27/15.
//  Copyright © 2015 AppsFoundation. All rights reserved.
//

import UIKit

class MenuItemsManager: NSObject {
    
    static let sharedManager = MenuItemsManager()
    
    private override init() {}
    
    // MARK: - Public Methods
    func loadData() -> [MenuItem] {
        let path = Bundle.main.path(forResource: "MenuItems", ofType: "plist")
        if let dataArray = NSArray(contentsOfFile: path!) {
            return constructMenuItemsFromArray(array: dataArray)
        } else {
            return [MenuItem]()
        }
    }
    
    // MARK: - Private Methods
    private func constructMenuItemsFromArray(array: NSArray) -> [MenuItem] {
        var resultItems = [MenuItem]()
        
        for object in array {
            let obj = object as! NSDictionary
            let name = obj["name"] as! String
            let ingredients = obj["ingredients"] as! String
            let image = obj["image"] as! String
            let price = obj["price"] as! String
            let discount = obj["discount"] as? String
            
            let loadedMenuItem = MenuItem(name: name, ingredients: ingredients, image: image, price: price, discount: discount)
            resultItems.append(loadedMenuItem)
        }
        return resultItems
    }
}
